amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/ExchangeMessageRouter.java

```diff
@@ -0,0 +1,451 @@
+            case Topic -> new TopicExchangeMessageRouter(exchange);
+            case Headers -> new HeadersExchangeMessageRouter(exchange);
+        };
+    }
+
+    static class FanoutExchangeMessageRouter extends ExchangeMessageRouter {
+
+        private final Set<Destination> destinationSet;
+
+        public FanoutExchangeMessageRouter(PersistentExchange exchange) {
+            super(exchange);
+            destinationSet = Sets.newConcurrentHashSet();
+        }
+
+        @Override
+        public synchronized void addBinding(String des, String desType, String routingKey,
+                                            Map<String, Object> arguments) {
+            destinationSet.add(new Destination(des, desType));
+        }
+
+        @Override
+        public synchronized void removeBinding(String des, String desType, String routingKey,
+                                               Map<String, Object> arguments) {
+            destinationSet.remove(new Destination(des, desType));
+        }
+
+        @Override
+        Set<Destination> getDestinations(String routingKey, Map<String, Object> headers) {
+            return destinationSet;
+        }
+
+    }
+
+    static class DirectExchangeMessageRouter extends ExchangeMessageRouter {
+
+        private final Map<String, Set<Destination>> destinationMap;
+
+        public DirectExchangeMessageRouter(PersistentExchange exchange) {
+            super(exchange);
+            destinationMap = new ConcurrentHashMap<>();
+        }
+
+        @Override
+        public synchronized void addBinding(String des, String desType, String routingKey,
+                                            Map<String, Object> arguments) {
+            destinationMap.computeIfAbsent(routingKey, k -> Sets.newConcurrentHashSet())
+                    .add(new Destination(des, desType));
+        }
+
+        @Override
+        public synchronized void removeBinding(String des, String desType, String routingKey,
+                                               Map<String, Object> arguments) {
+            destinationMap.computeIfPresent(routingKey, (k, v) -> {
+                v.remove(new Destination(des, desType));
+                if (v.isEmpty()) {
+                    return null;
+                }
+                return v;
+            });
+        }
+
+        @Override
+        Set<Destination> getDestinations(String routingKey, Map<String, Object> headers) {
+            return destinationMap.get(routingKey);
+        }
+    }
+
+    static class TopicExchangeMessageRouter extends ExchangeMessageRouter {
+
+        private final Map<Destination, TopicRoutingKeyParser> destinationMap;
+
+        public TopicExchangeMessageRouter(PersistentExchange exchange) {
+            super(exchange);
+            destinationMap = new ConcurrentHashMap<>();
+        }
+
+        static class TopicRoutingKeyParser {
+
+            Set<String> bindingKeys;
```

Comment:
The TopicRoutingKeyParser.bindingKeys needs to be initialized.

Reply:
fixed